vowpalwabbit/reductions/epsilon_decay.cc

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "epsilon_decay.h"
+#include "global_data.h"
+#include "learner.h"
+#include "memory.h"
+#include "vw.h"
+#include "label_type.h"
+#include "prediction_type.h"
+#include "model_utils.h"
+
+#include <string>
+#include <algorithm>
+#include <cmath>
+
+using namespace VW::config;
+using namespace VW::LEARNER;
+
+// namespace logger = VW::io::logger;
+
+namespace VW
+{
+namespace epsilon_decay
+{
+void epsilon_decay_score::update_bounds(float w, float r)
+{
+  update(w, r);
+
+  // update the lower bound
+  distributionally_robust::ScoredDual sd = this->chisq.recompute_duals();
+  _lower_bound = static_cast<float>(sd.first);
+}
+
+template <class ForwardIt>
+ForwardIt swap_models(ForwardIt first, ForwardIt n_first, ForwardIt end)
+{
+  for (; first != end; ++first, ++n_first) { std::iter_swap(first, n_first); }
+  return n_first;
+}
+
+template <class ForwardIt>
+void reset_models(ForwardIt first, ForwardIt end, parameters& weights, double epsilon_decay_alpha,
+    double epsilon_decay_tau, uint64_t model_count)
+{
+  uint64_t ppw = 1;
```

Comment:
What is ppw? Can you update name or add a comment

Reply:
params_per_weight -- changed